[SPARK-13842][PYSPARK] pyspark.sql.types.StructType accessor enhancements

shea-parkes opened the pull request:
## What changes were proposed in this pull request?

Expand the possible ways to interact with the contents of a `pyspark.sql.types.StructType` instance.
- Iterating a `StructType` will iterate its fields
  - `[field.name for field in my_structtype]`
- Indexing with a string will return a field by name
  - `my_structtype['my_field_name']`
- Indexing with an integer will return a field by position
  - `my_structtype[0]`
- Indexing with a slice will return a new `StructType` with just the chosen fields:
  - `my_structtype[1:3]`
- The length is the number of fields (should also provide "truthiness" for free)
  - `len(my_structtype) == 2`
## How was this patch tested?

Extended the unit test coverage in the accompanying `tests.py`.

shea-parkes committed:
Expand pyspark.sql.types.StructType tests.

Do this to cover the anticipated new APIs.

shea-parkes committed:
Expand pyspark.sql.types.StructType accessors.

Do this to support more of the Pythonic "magic" methods.

shea-parkes committed:
Dogfood new __iter__ in pyspark.sql.types.StructField.

Do this because the more direct syntax is now available.

shea-parkes committed:
Make pyspark.sql.types.StructType attributes lazier.

Do this so there is less chance for someone to get the stateful
attributes out of sync.

AmplabJenkins commented:
Can one of the admins verify this patch?

holdenk commented on python/pyspark/sql/types.py:
```diff
@@ -507,20 +503,52 @@ def add(self, field, data_type=None, nullable=True, metadata=None):
+    def __getitem__(self, key):
+        """Access fields by name or slice."""
+        if isinstance(key, str):
+            _dict_fields = {field.name: field for field in self}
```
it seems a bit expensive to compute this dictionary every time no?

skparkes replied:
I figured the performance of the metadata manipulation would be negligible as soon as you interacted with even a small dataset.

I could look into other lookup methods, but this seemed the most straightforward and pythonic.

skparkes replied:
A brief bit of googling suggests the only reasonable alternative would be a direct iteration with an early exit when you find a match.  I can switch to that if you prefer; just let me know.

holdenk replied:
So looking some more at the Scala implementation, this is done in such a way that the order of the fields returned is in the order of the fields in the schema  - it would be good to preserve this behaviour in the Python API for consistencies sake (and this implementation might I'm not super sure on Python slicing behaviour with dictionaries) but it would probably be good to add a test for this behaviour as well. What do you think?

skparkes replied:
The dictionary will only apply if they access via a `str`, so it can only return a single field.

It did just occur to me that using an actual slice object will return a `list` of `StructField` objects instead of a `StructType` object in this code currently.  I'll go fix that now.

skparkes replied:
There we go, I added another commit to correctly return `StructType` objects when accessing fields via a `slice`.  This was a quirk of Python I was only recently made aware of.  So in summary, this pull request proposes to:
  - Using a `str` key will return the named `StructField`: `my_structtype['my_structfield_name`]`
  - Using a single `int` key will return the `StructField` by index: `my_structtype[2]`
  - Using a `slice` object (e.g. `[2:3]`) will return a new `StructType` object with the chosen fields: `my_structtype[2:3]`

I'm happy to expand the tests and/or the documentation if you think any section could be clearer.

Thanks.

davies replied:
I think it's OK if this is not called for every rows. In that case, the index should be used for performance.

skparkes commented:
Discussed most of these changes ahead of time with @holdenk and @davies in the associated JIRA.  The last commit might be considered extraneous, but I thought it was related enough and a good idea so I went ahead and offered it up.  Easy enough to remove if people would prefer.

Ran just the `pyspark.sql.types.StructType` tests locally on a Python 3.5 setup.  Tried to be mindful of coding for python 2.7 compatibility (although I already miss type hints...).

Should be ready to review/discuss.

Thanks!

rxin commented:
Can you update the description to include the ways that are being expanded?

skparkes commented:
I updated the pull request description, but I also realize your comment could have meant the user documentation.  I'll spend a moment to see if I can find the appropriate location for that.

shea-parkes committed:
Expand pyspark.sql.types.StructType docstring.

Do this so the new accessors are better documented.

skparkes commented:
It appears the user documentation is nicely scraped from the docstrings, so I went ahead and expanded the docstring slightly (and added a couple `doctest`s).

Let me know if you intended something different.

shea-parkes committed:
Correctly slice StructTypes.

Do this because slicing containers in Python should return container
objects of the same type.

skparkes commented:
I updated the pull request description to reflect the fix for using the `slice` syntax.  It is my belief that this should offer a solid balance of functionality and the [principal of least astonishment](https://en.wikipedia.org/wiki/Principle_of_least_astonishment).  I'm still happy to revise this further (and even trim down the proposed changes).  Just let me know.

rxin commented:
cc @davies for review

davies commented on python/pyspark/sql/types.py:
```diff
@@ -507,20 +512,54 @@ def add(self, field, data_type=None, nullable=True, metadata=None):
                 data_type_f = data_type
             self.fields.append(StructField(field, data_type_f, nullable, metadata))
-            self.names.append(field)
-        self._needSerializeAnyField = any(f.needConversion() for f in self.fields)
```
This is used when unpickle, it's better to cache it.

davies commented on python/pyspark/sql/types.py:
```diff
@@ -457,13 +466,10 @@ def __init__(self, fields=None):
         """
         if not fields:
             self.fields = []
-            self.names = []
```
This is needed for performance (for serialization,  toInternal and fromInternal)

davies commented:
@skparkes These accessor look good, but have some concern about the performance of serialization, please don't remove the without a good reason.

shea-parkes committed:
Revert "Make pyspark.sql.types.StructType attributes lazier."

This reverts commit 831f03a3098d0e165703db6be0f4d76ef4893667.

skparkes commented:
Alright, I had all the `@property` work rolled into a single commit.  I've reverted it so only the accessor work should be left.  Let me know if you see anything else you think should be changed.

davies commented:
LGTM

SparkQA commented:
**[Test build #2831 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2831/consoleFull)** for PR 12251 at commit [`387cdfe`](https://github.com/apache/spark/commit/387cdfe0050f41f35710272e51593c484e5d0bfe).

SparkQA commented:
**[Test build #2831 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2831/consoleFull)** for PR 12251 at commit [`387cdfe`](https://github.com/apache/spark/commit/387cdfe0050f41f35710272e51593c484e5d0bfe).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

skparkes commented:
I apologize, I did not realize you were still targeting python 2.6 compatibility.  You guys really are gluttons for punishment.  I'll get a commit up today that should provide 2.6 compatibility.

davies commented:
@skparkes We had decided to drop Python 2.6 support, so we should run the tests with Python 2.7, will send a PR to fix that.

davies commented:
After reading the discussion again, it seems that we still need some work to drop Python 2.6. So go ahead to fix this for 2.6 please.

shea-parkes committed:
Make changes compatible with python 2.6.

Do this because Spark is still targeting python 2.6 compatibility.

skparkes commented:
I believe this pull request should now be compatible with python 2.6.  I did stand up a python 2.6 environment to run the pertinent tests locally and they now pass.

skparkes commented:
The final code looks a little silly without a dictionary comprehension now that I reflect on it.  Let me swap that over to just direct returning from the iteration.

shea-parkes committed:
Switch to returning field from inside loop.

Do this because without the dictionary comprehension syntax this was
looking very busy.

skparkes commented:
Okay, I made the switch over to the boring linear search method when accessing by field name.  I'm comfortable with the state of this pull request and believe it should pass tests on python 2.6 now.  Thanks.

SparkQA commented:
**[Test build #2838 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2838/consoleFull)** for PR 12251 at commit [`fc7898b`](https://github.com/apache/spark/commit/fc7898b0999aba15fc43465fddc116cea7c14f06).

SparkQA commented:
**[Test build #2838 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2838/consoleFull)** for PR 12251 at commit [`fc7898b`](https://github.com/apache/spark/commit/fc7898b0999aba15fc43465fddc116cea7c14f06).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

davies commented:
Merging this into master, thanks!

skparkes commented:
Thank you guys for working through this.
